node/router/src/lib.rs

```diff
@@ -251,6 +256,53 @@ impl<N: Network> Router<N> {
         !self.is_local_ip(ip) && !is_bogon_ip(ip.ip()) && !is_unspecified_or_broadcast_ip(ip.ip())
     }
 
+    /// Returns `true` if the message version is valid.
+    pub fn is_valid_message_version(&self, message_version: u32) -> bool {
+        // Fetch the latest message version.
+        let latest_message_version = Message::<N>::latest_message_version();
+
+        // Determine the minimum message version this node will accept, based on its role.
+        // - Provers always operate at the latest message version.
+        // - Validators and clients may accept older versions, depending on their current block height.
+        //
+        // Example scenario:
+        // - At block height `X`, the protocol upgrades to message version from `Y-1` to `Y`.
+        // - Client A upgrades and starts using message version `Y`.
+        // - Client B has not upgraded and still uses message version `Y-1`.
+        // - Until block `X`, they stay connected and can communicate.
+        // - After block `X`, Client A will reject messages from Client B.
+        let lowest_accepted_message_version = match self.node_type {
+            // Provers should always use the latest version.
+            NodeType::Prover => Message::<N>::latest_message_version(),
+            // Validators and clients accept messages from lower version based on the migration height.
+            NodeType::Validator | NodeType::Client => {
+                let latest_block_height = self.ledger.latest_block_height();
+                let versions = Message::<N>::VERSIONS;
+                // TODO (raychu86): Refactor `consensus_config_value` to support generic constant types.
+                // Determine the message version to use.
+                N::CONSENSUS_VERSION(latest_block_height).map_or(latest_message_version, |seek_version| {
+                    // Search the consensus value for the specified version.
+                    match versions.binary_search_by(|(version, _)| version.cmp(&seek_version)) {
+                        // If a value was found for this consensus version, return it.
+                        Ok(index) => versions[index].1,
+                        Err(index) => {
+                            // If the specified version was not found exactly, determine whether to return an appropriate value anyway.
+                            match index {
+                                // This constant is not yet in effect at this consensus version.
+                                0 => Message::<N>::latest_message_version(),
```

Comment:
If I understand correctly, at block heights `0..N::CONSENSUS_HEIGHT(ConsensusVersion::V4` this will return 17 (due to `latest_message_version`, then at `ConsensusVersion::V4` jump to 16, then at `ConsensusVersion::V5` jump back to 17.

I think one way to fix this would be to use `ConsensusVersion::V1, 16` in the `VERSIONS` constant, so that it'll start at 16.

We should test with the test feature [once that PR is ready](https://github.com/ProvableHQ/snarkOS/pull/3572), so that we can observe the behaviour at low block heights. CC @kpandl

Reply:
That't not exactly correct. 

You will be using `17` from `0..N::CONSENSUS_HEIGHT(ConsensusVersion::V5)`. With this change, you will ALWAYS be using the latest version number at 17. However, you are able to accept connections from nodes using version 16 up until `N::CONSENSUS_HEIGHT(ConsensusVersion::V6)`

Reply:
The new logic in [3eca0df](https://github.com/ProvableHQ/snarkOS/pull/3596/commits/3eca0dfa4df035e58f66ebc482c3bcc94dd3119a) should address this